src/StimulusBundle/src/Dto/StimulusAttributes.php

```diff
@@ -107,54 +107,43 @@ public function addAttribute(string $name, string $value): void
 
     public function __toString(): string
     {
-        $controllers = array_map(function (string $controllerName): string {
-            return $this->escapeAsHtmlAttr($controllerName);
-        }, $this->controllers);
+        $attributes = [];
 
-        // done separately so we can escape, but avoid escaping ->
-        $actions = array_map(function (array $actionData): string {
-            $controllerName = $this->escapeAsHtmlAttr($actionData['controllerName']);
-            $actionName = $this->escapeAsHtmlAttr($actionData['actionName']);
-            $eventName = $actionData['eventName'];
+        if ($this->controllers) {
+            $attributes[] = 'data-controller="'.$this->escape(implode(' ', $this->controllers), 'html').'"';
+        }
 
-            $action = $controllerName.'#'.$actionName;
-            if (null !== $eventName) {
-                $action = $this->escapeAsHtmlAttr($eventName).'->'.$action;
-            }
+        if ($this->actions) {
+            // done separately so we can escape, but avoid escaping ->
+            $actions = array_map(function (array $actionData): string {
+                $controllerName = $actionData['controllerName'];
+                $actionName = $actionData['actionName'];
+                $eventName = $actionData['eventName'];
 
-            return $action;
-        }, $this->actions);
-
-        $targets = [];
-        foreach ($this->targets as $key => $targetNamesString) {
-            $targetNames = explode(' ', $targetNamesString);
-            $targets[$key] = implode(' ', array_map(function (string $targetName): string {
-                return $this->escapeAsHtmlAttr($targetName);
-            }, $targetNames));
-        }
+                $action = $this->escape($controllerName.'#'.$actionName, 'html');
+                if (null !== $eventName) {
+                    $action = $this->escape($eventName, 'html').'->'.$action;
+                }
 
-        $attributes = [];
+                return $action;
+            }, $this->actions);
 
-        if ($controllers) {
-            $attributes[] = \sprintf('data-controller="%s"', implode(' ', $controllers));
+            $attributes[] = 'data-action="'.implode(' ', $actions).'"';
         }
 
-        if ($actions) {
-            $attributes[] = \sprintf('data-action="%s"', implode(' ', $actions));
+        if ($this->targets) {
+            $attributes[] = implode(' ', array_map(function (string $key, string $value): string {
+                return $this->escape($key, 'html_attr').'="'.$this->escape($value, 'html').'"';
+            }, array_keys($this->targets), $this->targets));
```

Comment:
foreach FTW, seriously :)
```suggestion
            foreach ($this->targets as $k => $v) {
                $attributes[] = $this->escape($k, 'html_attr').'="'.$this->escape($v).'"';
            }
```

Reply:
Done